File: libs/crypto.mjs
import crypto from 'crypto'
import log from './console.mjs'

export const ImportKey = (jwk, name = 'ECDH') => {
    return crypto.subtle.importKey(
        'jwk',
        jwk,
        {
            name,
            namedCurve: jwk.crv
        },
        true,
        jwk.key_ops
    )
}

export let GlobalJWT = {}

export const GetPrivateKey = (jwk) => jwk.d

let publicKeyCache = ''

export const GetPublicKey = (jwk) => {
    if (!publicKeyCache) {
        publicKeyCache = base64_to_base64url(buffer_to_base64(concatBuffer(new Uint8Array([4]).buffer, base64_to_buffer(base64url_to_base64(jwk.x)), base64_to_buffer(base64url_to_base64(jwk.y)))))
    }
    return publicKeyCache
}

export const BuildJWT = async (vapidObject = {}, aud = '') => {
    const now = Date.now()

    if (GlobalJWT[aud] && GlobalJWT[aud].content && GlobalJWT[aud].expire > now) {
        return GlobalJWT[aud].content
    }

    if (!(vapidObject.key && aud && vapidObject.sub)) {
        return ''
    }
    try {
        const info = {
            typ: 'JWT',
            alg: 'ES256'
        }
        const data = {
            aud,
            exp: Math.floor(now / 1000) + 60 * 60,
            sub: vapidObject.sub
        }
        const unsignedToken = base64_to_base64url(btoa(JSON.stringify(info))) + '.' + base64_to_base64url(btoa(JSON.stringify(data)))

        GlobalJWT[aud] = {
            content: unsignedToken + '.' + base64_to_base64url(buffer_to_base64(await Sign(vapidObject.key, new TextEncoder().encode(unsignedToken)))),
            expire: now + 30 * 60 * 1000
        }

        return GlobalJWT[aud].content
    } catch (e) {
        log.error(e)
        return ''
    }
}

export const Sign = async (jwk = {}, payload = new Uint8Array()) => {
    const key = await ImportKey(jwk, 'ECDSA')
    return await crypto.subtle.sign(
        {
            name: 'ECDSA',
            hash: {
                name: 'SHA-256'
            }
        },
        key,
        payload
    )
}

//https://stackoverflow.com/questions/21797299/convert-base64-string-to-arraybuffer
export const base64_to_buffer = (base64) => {
    let binaryString = atob(base64)
    let bytes = new Uint8Array(binaryString.length)
    for (let i = 0; i < binaryString.length; i++) {
        bytes[i] = binaryString.charCodeAt(i)
    }
    return bytes.buffer
}
//https://stackoverflow.com/questions/56846930/how-to-convert-raw-representations-of-ecdh-key-pair-into-a-json-web-key
export const hex_to_uintarray = (hex) => {
    const a = []
    for (let i = 0, len = hex.length; i < len; i += 2) {
        a.push(parseInt(hex.substr(i, 2), 16))
    }
    return new Uint8Array(a)
}
export const buffer_to_base64 = (buf) => {
    let binary = ''
    const bytes = new Uint8Array(buf)
    for (var i = 0; i < bytes.byteLength; i++) {
        binary += String.fromCharCode(bytes[i])
    }
    return btoa(binary)
}
export const base64_to_base64url = (base64) => base64.replaceAll('/', '_').replaceAll('+', '-').replaceAll('=', '')
export const base64url_to_base64 = (base64url) => base64url.replaceAll('_', '/').replaceAll('-', '+')
//https://stackoverflow.com/questions/40031688/javascript-arraybuffer-to-hex
export const buffer_to_hex = (buffer) => {
    // buffer is an ArrayBuffer
    return [...new Uint8Array(buffer)].map((x) => x.toString(16).padStart(2, '0')).join('')
}
//https://gist.github.com/72lions/4528834
export const concatBuffer = (...buffer) => {
    const length = buffer.reduce((acc, cur) => acc + cur.byteLength, 0)
    let tmp = new Uint8Array(length)
    buffer.reduce((acc, cur) => {
        tmp.set(new Uint8Array(cur), acc)
        return acc + cur.byteLength
    }, 0)
    return tmp
}

export const ECDH = async (publicKey, privateKey) => {
    const ecdh_secret_CryptoKey = await crypto.subtle.deriveKey(
        {
            name: 'ECDH',
            public: publicKey
        },
        privateKey,
        { name: 'AES-GCM', length: 256 },
        true,
        ['encrypt', 'decrypt']
    )
    const ecdh_secret = await crypto.subtle.exportKey('raw', ecdh_secret_CryptoKey)
    return ecdh_secret
}

export const HmacSHA256 = async (key, data) => {
    const keyData = await crypto.subtle.importKey('raw', key, { name: 'HMAC', hash: 'SHA-256' }, false, ['sign', 'verify'])
    return new Uint8Array(await crypto.subtle.sign('HMAC', keyData, data))
}
export const HKDF = async (key, ikm, info, length = -1) => {
    const tmpKey = await HmacSHA256(key, ikm)
    return (await HmacSHA256(tmpKey, info)).slice(0, length < 0 ? undefined : length)
}

export const GetAESGCMNonceAndCekAndContent = async (subscriptionPublicKeyStr, auth, eccKeyData, salt) => {
    const subscriptionPublicKeyBuffer = new Uint8Array(base64_to_buffer(base64url_to_base64(subscriptionPublicKeyStr)))
    const subscriptionPublicKey = await crypto.subtle.importKey(
        'raw',
        subscriptionPublicKeyBuffer,
        {
            name: 'ECDH',
            namedCurve: 'P-256'
        },
        true,
        []
    )
    const ecdh_secret = await ECDH(subscriptionPublicKey, eccKeyData.privateKey)

    const publishPublicKeyBuffer = await crypto.subtle.exportKey('raw', eccKeyData.publicKey)

    const auth_secret = base64_to_buffer(base64url_to_base64(auth))

    const context = concatBuffer(new TextEncoder().encode('P-256\0'), new Uint8Array([0, 65]), subscriptionPublicKeyBuffer, new Uint8Array([0, 65]), publishPublicKeyBuffer)

    const auth_info = new TextEncoder().encode('Content-Encoding: auth\0')
    const PRK_combine = await HmacSHA256(auth_secret, ecdh_secret)
    const IKM = await HmacSHA256(PRK_combine, concatBuffer(auth_info, new Uint8Array([1])))

    const PRK = await HmacSHA256(salt, IKM)
    const cek_info = concatBuffer(new TextEncoder().encode('Content-Encoding: aesgcm\0'), context)
    let cek = (await HmacSHA256(PRK, concatBuffer(cek_info, new Uint8Array([1])))).slice(0, 16)
    const nonce_info = concatBuffer(new TextEncoder().encode('Content-Encoding: nonce\0'), context)
    let nonce = (await HmacSHA256(PRK, concatBuffer(nonce_info, new Uint8Array([1])))).slice(0, 12)
    return { nonce, cek, context }
}

export const GetAES128GCMNonceAndCekAndContent = async (subscriptionPublicKeyStr, auth, eccKeyData, salt) => {
    const subscriptionPublicKeyBuffer = new Uint8Array(base64_to_buffer(base64url_to_base64(subscriptionPublicKeyStr)))
    const subscriptionPublicKey = await crypto.subtle.importKey(
        'raw',
        subscriptionPublicKeyBuffer,
        {
            name: 'ECDH',
            namedCurve: 'P-256'
        },
        true,
        []
    )
    const ecdh_secret = await ECDH(subscriptionPublicKey, eccKeyData.privateKey)

    const publishPublicKeyBuffer = await crypto.subtle.exportKey('raw', eccKeyData.publicKey)

    const auth_secret = base64_to_buffer(base64url_to_base64(auth))

    const key_info = concatBuffer(new TextEncoder('utf-8').encode('WebPush: info\0'), subscriptionPublicKeyBuffer, publishPublicKeyBuffer)

    const PRK_key = await HmacSHA256(auth_secret, ecdh_secret)

    let IKM = await HmacSHA256(PRK_key, concatBuffer(key_info, new Uint8Array([1]).buffer))
    let PRK = await HmacSHA256(salt, IKM)
    let cek_info = new TextEncoder('utf-8').encode('Content-Encoding: aes128gcm\0')
    let contentEncryptionKey = (await HmacSHA256(PRK, concatBuffer(cek_info, new Uint8Array([1]).buffer))).slice(0, 16)
    let nonce_info = new TextEncoder('utf-8').encode('Content-Encoding: nonce\0')
    let nonce = (await HmacSHA256(PRK, concatBuffer(nonce_info, new Uint8Array([1]).buffer))).slice(0, 12)

    return { nonce, cek: contentEncryptionKey, context: key_info }
}

export const Encrypt = async (nonce, contentEncryptionKey, content, encoding = 'aesgcm') => {
    const cek = await crypto.subtle.importKey('raw', contentEncryptionKey, 'AES-GCM', true, ['encrypt', 'decrypt'])
    let payload
    if (encoding === 'aes128gcm') {
        payload = concatBuffer(content, new Uint8Array([2]))
    } else {
        payload = concatBuffer(new Uint8Array([0, 0]), content)
    }
    let encodedBuffer = await crypto.subtle.encrypt({ name: 'AES-GCM', iv: nonce }, cek, payload)
    return encodedBuffer
}

File: functions/crypto.go
package functions

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdh"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"math/big"
	"strings"
	"sync"
	"time"

	"github.com/golang-jwt/jwt"
)

type JWT struct {
	Content string
	Expire  int64
}

// ImportKey imports a JWK as an ECDSA private key.
func ImportKey(_privateKey string) (*ecdsa.PrivateKey, error) {
	dBytes, err := base64.RawURLEncoding.DecodeString(_privateKey)
	if err != nil {
		return nil, err
	}

	curve := elliptic.P256()

	d := new(big.Int).SetBytes(dBytes)

	privateKey := &ecdsa.PrivateKey{
		PublicKey: ecdsa.PublicKey{
			Curve: curve,
			X:     nil,
			Y:     nil,
		},
		D: d,
	}

	privateKey.X, privateKey.Y = curve.ScalarBaseMult(d.Bytes())

	return privateKey, nil
}

// GetPrivateKey retrieves the private key from JWK.
func GetPrivateKey(privateKey *ecdsa.PrivateKey) string {
	return strings.ReplaceAll(base64.RawURLEncoding.EncodeToString(privateKey.D.Bytes()), "=", "")
}

// GetPublicKey retrieves or caches the public key derived from JWK.
func GetPublicKey(privateKey *ecdsa.PrivateKey) string {
	pk, _ := privateKey.PublicKey.ECDH()
	return strings.ReplaceAll(base64.RawURLEncoding.EncodeToString(pk.Bytes()), "=", "")
}

var GlobalJWT sync.Map

type GlobalJWTContent struct {
	Content string
	Expire  int64
}

// BuildJWT builds a JWT token using the provided VAPID object and audience.
func BuildJWT(privateKey *ecdsa.PrivateKey, aud string, sub string) (string, error) {
	now := Now
	if jwt, exists := GlobalJWT.Load(aud); exists {
		j := jwt.(GlobalJWTContent)
		if j.Content != "" && j.Expire > now.UnixMilli() {
			return j.Content, nil
		}
	}

	if aud == "" {
		return "", errors.New("missing audience")
	}

	// Create the Claims
	claims := jwt.StandardClaims{
		Subject: sub,
		// NotBefore: now.Unix(),
		// IssuedAt:  now.Unix(),
		ExpiresAt: now.Add(time.Hour).Unix(),
		Audience:  aud,
	}

	token := jwt.NewWithClaims(jwt.SigningMethodES256, claims)
	ss, err := token.SignedString(privateKey)
	if err == nil {
		GlobalJWT.Store(aud, GlobalJWTContent{
			Content: ss,
			Expire:  now.Add(time.Minute * 30).UnixMilli(),
		})
	}

	return ss, err
}

// Sign signs the payload with the provided JWK key using ECDSA.
func Sign(privateKey *ecdsa.PrivateKey, payload []byte) ([]byte, error) {
	hash := sha256.Sum256(payload)
	r, s, err := ecdsa.Sign(rand.Reader, privateKey, hash[:])
	if err != nil {
		return nil, err
	}

	return append(r.Bytes(), s.Bytes()...), nil
}

func GenHMAC256(ciphertext, key []byte) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(ciphertext))
	hmac := mac.Sum(nil)
	return hmac
}

func GetAESGCMNonceAndCekAndContent(subscriptionPublicKey *ecdh.PublicKey, auth []byte, privateKey *ecdsa.PrivateKey, salt []byte) ([]byte, []byte, []byte) {
	subscriptionPublicKeyBuffer := subscriptionPublicKey.Bytes()
	ecdhPrivateKey, _ := privateKey.ECDH()
	ecdh_secret, _ := ecdhPrivateKey.ECDH(subscriptionPublicKey)
	publishPublicKeyBuffer := ConcatBuffer([]byte{4}, privateKey.PublicKey.X.Bytes(), privateKey.PublicKey.Y.Bytes())

	auth_secret := auth

	context := []byte("P-256\x00")
	context = append(context, []byte{0, 65}...)
	context = append(context, subscriptionPublicKeyBuffer...)
	context = append(context, []byte{0, 65}...)
	context = append(context, publishPublicKeyBuffer...)

	auth_info := []byte("Content-Encoding: auth\x00")
	PRK_combine := GenHMAC256(ecdh_secret, auth_secret)
	IKM := GenHMAC256(append(auth_info, 0x01), PRK_combine)

	PRK := GenHMAC256(IKM, salt)

	cek_info := []byte("Content-Encoding: aesgcm\x00")
	cek_info = append(cek_info, context...)
	cek := GenHMAC256(append(cek_info, 0x01), PRK)[0:16]

	nonce_info := []byte("Content-Encoding: nonce\x00")
	nonce_info = append(nonce_info, context...)
	nonce := GenHMAC256(append(nonce_info, 0x01), PRK)[0:12]

	return nonce, cek, context
}

func GetAES128GCMNonceAndCekAndContent(subscriptionPublicKey *ecdh.PublicKey, auth []byte, privateKey *ecdsa.PrivateKey, salt []byte) ([]byte, []byte, []byte) {
	subscriptionPublicKeyBuffer := subscriptionPublicKey.Bytes()

	ecdhPrivateKey, _ := privateKey.ECDH()
	ecdh_secret, _ := ecdhPrivateKey.ECDH(subscriptionPublicKey)
	publishPublicKeyBuffer := ConcatBuffer([]byte{4}, privateKey.PublicKey.X.Bytes(), privateKey.PublicKey.Y.Bytes())

	auth_secret := auth

	key_info := []byte("WebPush: info\x00")
	key_info = append(key_info, subscriptionPublicKeyBuffer...)
	key_info = append(key_info, publishPublicKeyBuffer...)

	PRK_key := GenHMAC256(ecdh_secret, auth_secret)
	IKM := GenHMAC256(append(key_info, 0x01), PRK_key)

	PRK := GenHMAC256(IKM, salt)

	cek_info := []byte("Content-Encoding: aes128gcm\x00")
	cek := GenHMAC256(append(cek_info, 0x01), PRK)[0:16]

	nonce_info := []byte("Content-Encoding: nonce\x00")
	nonce := GenHMAC256(append(nonce_info, 0x01), PRK)[0:12]

	return nonce, cek, key_info
}

func Encrypt(nonce, contentEncryptionKey, content []byte, encoding string) []byte {

	block, err := aes.NewCipher(contentEncryptionKey)
	if err != nil {
		return nil
	}
	aead, err := cipher.NewGCM(block)
	if err != nil {
		return nil
	}

	payload := []byte{}

	if encoding == "aes128gcm" {
		payload = append(payload, 0x02)
	} else {
		tmp := []byte("\x00\x00")
		payload = append(tmp, payload...)
	}

	ciphertext := aead.Seal(nonce, nonce, payload, nil)

	return ciphertext

}
